### PR TITLE
Expand documentation on the role of API and our testing approach

### PR DESCRIPTION
**Pull request by Geod24**
CC @denizzzka : Does this documentation make sense ?

### CODE REVIEW

File: source/agora/common/API.d [hunk 1]
```diff
@@ -1,6 +1,27 @@
...
+    on the data it receives, and can forward that data to other nodes.
+    A validator is a full node which participate in consensus.
+
+    An `API` is used as an interface to communicate with a node.
+    As such, a class that implements `API` exists (in `agora.node.Node`),
+    and in order to communicate with other nodes, it holds an `API` for
+    each of those nodes.
+
+    `API` are defined as D interfaces, following what is done in Vibe.d.
+    Those interfaces can be read by a generator to build a client.
```
Comment (denizzzka) [inline source/agora/common/API.d]: ...to build client and server ?

**Comment by denizzzka** [comment]
> Does this documentation make sense ?

Yes, it was useful, I learned something

File: source/agora/common/API.d (hunk 1 above)
Comment (Geod24) [reply source/agora/common/API.d]: Amended the doc to mention server side

**Comment by Geod24** [comment]
💚

File: source/agora/common/API.d [hunk 2]
```diff
@@ -1,6 +1,28 @@
 /*******************************************************************************
 
-    Contains the REST API interface.
+    Definitions of the nodes (full node & validator) APIs
+
+    Two kinds of nodes exist: full nodes, and validators.
+    A full node follows the network as a passive actor, but do validation
```
Comment (AndrejMitrovic) [inline source/agora/common/API.d]:
```suggestion
    A full node follows the network as a passive actor, but does validation
```

File: source/agora/common/API.d [hunk 3]
```diff
@@ -1,6 +1,28 @@
...
+    on the data it receives, and can forward that data to other nodes.
+    A validator is a full node which participate in consensus.
```
Comment (AndrejMitrovic) [inline source/agora/common/API.d]:
```suggestion
    A validator is a full node which participates in consensus.
```

File: source/agora/test/README.md [hunk 4]
```diff
@@ -0,0 +1,26 @@
+# Agora test folder
+
+This folder contains network tests for Agora.
+Network tests are the second layer of tests in our 3 layers test architecture,
+the first one being unit tests, and the last one integration tests.
+They are implemented in `unittest` blocks but are more involved than unit tests.
+
+Network tests involve the `localrest` library, which simulate a network situation,
```
Comment (AndrejMitrovic) [inline source/agora/test/README.md]:
```suggestion
Network tests involve the `localrest` library, which simulates a network situation,
```

File: source/agora/test/README.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+and a task manager. The core idea is that localrest spawns a thread per "remote node",
+and allow a client, living in the main thread, to make requests to one or more
```
Comment (AndrejMitrovic) [inline source/agora/test/README.md]:
```suggestion
and allows a client, living in the main thread, to make requests to one or more
```

File: source/agora/test/README.md [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+of those remote nodes. They then communicate with each other via message passing.
+This approach is surprisingly efficient to model a node's behavior,
+and also forces a design where the network code is decoupled from the business code.
+
+When adding a test, a good starting point is to look at `agora.test.Base`,
+a module that provides basic functionality to write tests.
+Note that this approach requires careful design of the business code.
+For example, when writing tasks, one have to use the `TaskManager`,
```
Comment (AndrejMitrovic) [inline source/agora/test/README.md]:
```suggestion
For example, when writing tasks, one has to use the `TaskManager`,
```

**Comment by AndrejMitrovic** [comment]
Nits, LGTM.

**Comment by Geod24** [comment]
> Nits, LGTM.

I see a pattern

**Review by AndrejMitrovic — APPROVED** [review]

**Pull request merged**